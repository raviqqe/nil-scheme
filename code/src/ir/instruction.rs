use crate::Operand;
#[cfg(feature = "alloc")]
use alloc::vec::Vec;
use core::fmt::{self, Display, Formatter};

#[derive(Debug, Eq, PartialEq)]
pub enum Instruction {
    Call(Operand),
    Set(Operand),
    Get(Operand),
    Constant(Operand),
    #[cfg(feature = "alloc")]
    If(Vec<Instruction>),
    #[cfg(feature = "alloc")]
    Closure(u64, Vec<Instruction>),
    Skip(u64),
}

impl Instruction {
    pub const CALL: u8 = 0;
    pub const SET: u8 = 1;
    pub const GET: u8 = 2;
    pub const CONSTANT: u8 = 3;
    pub const IF: u8 = 4;
    pub const CLOSURE: u8 = 5;
    pub const SKIP: u8 = 6;
}

pub(crate) struct DisplayInstruction<'a> {
    instruction: &'a Instruction,
    indent: usize,
}

impl<'a> DisplayInstruction<'a> {
    pub fn new(instruction: &'a Instruction, indent: usize) -> Self {
        Self {
            instruction,
            indent,
        }
    }
}

impl<'a> Display for DisplayInstruction<'a> {
    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
        let indent = self.indent + 1;

        match self.instruction {
            Instruction::Call(operand) => write!(formatter, "call {}", operand),
            Instruction::Set(operand) => write!(formatter, "set {}", operand),
            Instruction::Get(operand) => write!(formatter, "get {}", operand),
            Instruction::Constant(operand) => write!(formatter, "constant {}", operand),
            #[cfg(feature = "alloc")]
            Instruction::If(instructions) => {
                writeln!(formatter, "if")?;
                write!(
                    formatter,
                    "{}",
                    DisplayInstructionList::new(instructions, indent)
                )
            }
            #[cfg(feature = "alloc")]
            Instruction::Closure(arity, instructions) => {
                writeln!(formatter, "closure {}", arity)?;
                write!(
                    formatter,
                    "{}",
                    DisplayInstructionList::new(instructions, indent)
                )
            }
            Instruction::Skip(count) => write!(formatter, "skip {}", count),
        }
    }
}

pub(crate) struct DisplayInstructionList<'a> {
    instructions: &'a [Instruction],
    indent: usize,
}

impl<'a> DisplayInstructionList<'a> {
    pub fn new(instructions: &'a [Instruction], indent: usize) -> Self {
        Self {
            instructions,
            indent,
        }
    }
}

impl<'a> Display for DisplayInstructionList<'a> {
    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
        for (index, instruction) in self.instructions.iter().enumerate() {
            for _ in 0..self.indent {
                write!(formatter, "  ")?
            }

            write!(
                formatter,
                "{}",
                DisplayInstruction::new(instruction, self.indent)
            )?;

            if index < self.instructions.len() - 1 {
                writeln!(formatter)?;
            }
        }

        Ok(())
    }
}
