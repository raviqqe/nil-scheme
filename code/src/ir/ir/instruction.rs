use crate::Operand;
use alloc::vec::Vec;

#[derive(Debug)]
pub enum Instruction {
    Apply(Operand, bool),
    Set(Operand),
    Get(Operand),
    Constant(u64),
    If(Vec<Instruction>, Vec<Instruction>),
}

impl Instruction {
    pub const APPLY_TAIL: u8 = 0;
    pub const APPLY: u8 = 1;
    pub const SET: u8 = 2;
    pub const GET: u8 = 3;
    pub const CONSTANT: u8 = 4;
    pub const IF: u8 = 5;
}
