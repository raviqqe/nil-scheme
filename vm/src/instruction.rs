#[derive(Clone, Copy, Debug)]
pub enum Instruction {
    Call,
    Set,
    Get,
    Constant,
    If,
    Halt,
}

impl Instruction {
    pub const CALL: u8 = Self::Call as _;
    pub const SET: u8 = Self::Set as _;
    pub const GET: u8 = Self::Get as _;
    pub const CONSTANT: u8 = Self::Constant as _;
    pub const IF: u8 = Self::If as _;
    pub const HALT: u8 = Self::Halt as _;
}
