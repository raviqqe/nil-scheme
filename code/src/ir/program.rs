use super::instruction::{DisplayInstructionList, Instruction};
use alloc::{string::String, vec::Vec};
use core::fmt::{self, Display, Formatter};

#[derive(Debug, Eq, PartialEq)]
pub struct Program {
    symbols: Vec<String>,
    instructions: Vec<Instruction>,
}

impl Program {
    pub fn new(symbols: Vec<String>, instructions: Vec<Instruction>) -> Self {
        Self {
            symbols,
            instructions,
        }
    }

    pub fn symbols(&self) -> &[String] {
        &self.symbols
    }

    pub fn instructions(&self) -> &[Instruction] {
        &self.instructions
    }
}

impl Display for Program {
    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
        writeln!(formatter, "symbols:")?;

        for symbol in &self.symbols {
            writeln!(formatter, "- {}", symbol)?;
        }

        write!(formatter, "instructions:")?;
        write!(
            formatter,
            "{}",
            DisplayInstructionList::new(&self.instructions, 0)
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Operand;
    use insta::assert_display_snapshot;
    use std::{format, vec};

    #[test]
    fn display_symbols() {
        assert_display_snapshot!(Program::new(vec!["foo".into(), "bar".into()], vec![],));
    }

    #[test]
    fn display_if() {
        assert_display_snapshot!(Program::new(
            vec![],
            vec![Instruction::If(vec![Instruction::Constant(
                Operand::Integer(42)
            )])],
        ));
    }

    #[test]
    fn display_closure() {
        assert_display_snapshot!(Program::new(
            vec![],
            vec![Instruction::Close(
                42,
                vec![Instruction::Constant(Operand::Integer(2045))],
            )],
        ));
    }

    #[test]
    fn display_closure_with_if() {
        assert_display_snapshot!(Program::new(
            vec![],
            vec![
                Instruction::Constant(Operand::Integer(0)),
                Instruction::Constant(Operand::Integer(1)),
                Instruction::Close(
                    42,
                    vec![
                        Instruction::Constant(Operand::Integer(2)),
                        Instruction::If(vec![
                            Instruction::Constant(Operand::Integer(3)),
                            Instruction::Constant(Operand::Integer(4)),
                        ]),
                        Instruction::Constant(Operand::Integer(5)),
                        Instruction::Constant(Operand::Integer(6)),
                    ],
                ),
                Instruction::Constant(Operand::Integer(7)),
                Instruction::Constant(Operand::Integer(8)),
            ],
        ));
    }
}
