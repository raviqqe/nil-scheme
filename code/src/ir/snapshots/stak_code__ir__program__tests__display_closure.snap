---
source: code/src/ir/program.rs
expression: "Program::new(vec![],\n    vec![Instruction ::\n        Closure(42, vec! [Instruction :: Call(Operand :: Integer(0))],)])"
---
symbols: []
instructions:
closure 42
  call i0
